test-tests/shared/src/test/scala/zio/test/environment/ClockSpec.scala

```diff
@@ -0,0 +1,156 @@
+package zio.test.environment
+
+import java.time.ZoneId
+import java.util.concurrent.TimeUnit
+
+import zio._
+import zio.clock._
+import zio.duration.Duration._
+import zio.duration._
+import zio.test.Assertion._
+import zio.test.TestAspect._
+import zio.test._
+import zio.test.environment.TestClock._
+
+object ClockSpec
+    extends ZIOBaseSpec(
+      suite("ClockSpec")(
+        testM("sleep does not require passage of clock time") {
+          for {
+            latch <- Promise.make[Nothing, Unit]
```

Comment:
I would use `TestClock.setTime(0.hours)` here. I think the reason you were having issues with some of these tests is that the testing framework provides a separate copy of the `TestEnvironment` to each test, but that environment is shared between successive invocations of the same test via `nonFlaky`. So like in this test on the first run the clock time will start at 0 hours and then you will sleep until the time is 10 hours. But on the second run the clock time will start at 11 hours so now an effect that sleeps for 10 hours will immediately run.

I think if you generally use the `TestClock` in the `TestEnvironment` and set the time to zero as your first action in any test where you are using `nonFlaky` that should get you to a pretty good place.

Reply:
>... the testing framework provides a separate copy of the TestEnvironment to each test, but that environment is shared between successive invocations of the same test via nonFlaky.

This is handy to know. I'll add a note to the aspect doco. seems like a good case for `@@ after(setTime(0.hours))`

> So like in this test on the first run the clock time will start at 0 hours and then you will sleep until the time is 10 hours. But on the second run the clock time will start at 11 hours so now an effect that sleeps for 10 hours will immediately run.

I don't think this example is quite right but I get the point :) (second run the clock starts at `11.hours`, then `sleep(10.hours)` sleeps to `21.hours`).

Reply:
@adamgfraser  I'm looking at "sleep correctly handles multiple sleeps" and my confusion is increasing... my (now improved) understanding of `nonFlaky` this is it's effectively checking the assertion over and over in the same environment, with the clock (in this particular case) increasing in 2 hr increments. Tasks are scheduled to start on odd hours, time is always even. The latches prevent the creations of race conditions caused by swapping fibers that the documentation mentions. The latches and ref are new for each iteration. Nothing seems dependent on the absolute time (sleep & adjust are both relative to the current time) so resetting the clock should have no effect here, yet this is still flaky without ` @@ after(setTime(0.hours))`

Is there some way of debugging in ZIO? I'm really missing being able to pause at arbitrary locations and check state.
```
for {                                    // time is 0 hours
  latch1 <- Promise.make[Nothing, Unit]  // new latch
  latch2 <- Promise.make[Nothing, Unit]  // new latch
  ref    <- Ref.make("")                 // new ref
                                         // set up fiber 1, delayed till 3.hours
  _      <- (sleep(3.hours) *> ref.update(_ + "World!") *> latch2.succeed(())).fork
                                         // set up fiber 2, delayed till 1.hour
  _      <- (sleep(1.hours) *> ref.update(_ + "Hello, ") *> latch1.succeed(())).fork 
  _      <- adjust(2.hours)              // adjust time to 2.hours, starts fiber 2
  _      <- latch1.await                 // wait for fiber 2 to finish, ref should now be "Hello, "
  _      <- adjust(2.hours)              // adjust time to 4.hours, starts fiber 1
  _      <- latch2.await                 // wait for fiber 1 to finish, ref should now be "Hello, World!"
  result <- ref.get
} yield assert(result, equalTo("Hello, World!"))
for {                                    // time is still 4.hours
  latch1 <- Promise.make[Nothing, Unit]  // new latch
  latch2 <- Promise.make[Nothing, Unit]  // new latch
  ref    <- Ref.make("")                 // new ref
                                         // set up fiber 3, delayed till 7.hours
  _      <- (sleep(3.hours) *> ref.update(_ + "World!") *> latch2.succeed(())).fork
                                         // set up fiber 4, delayed till 5.hours
  _      <- (sleep(1.hours) *> ref.update(_ + "Hello, ") *> latch1.succeed(())).fork
  _      <- adjust(2.hours)              // adjust time to 6.hours, starts fiber 4
  _      <- latch1.await                 // wait for fiber 4 to finish, ref should now be "Hello, "
  _      <- adjust(2.hours)              // adjust time to 8.hours, starts fiber 3
  _      <- latch2.await                 // wait for fiber 3 to finish, ref should now be "Hello, World!"
  result <- ref.get
} yield assert(result, equalTo("Hello, World!"))
//rinse, repeat 98 more times
```